### PR TITLE
Add `gasPrice` to 0x transactions

### PR DESCRIPTION
**Pull request by abandeali1**
## Description

- Adds a `gasPrice` field to 0x transactions. Transactions can now only be executed if the Ethereum tx `gasPrice` matches that of the 0x tx.
- Breaks out all of the validation in `executeTransaction` into an `_assertExecutableTransaction` function.
- Adds to and refactors #2069 to reduce some code duplication.
- Introduces a new `TransactionInvalidContextError` that replaces the previous reentrancy error.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/exchange/contracts/src/MixinTransactions.sol [hunk 1]
```diff
@@ -96,6 +140,16 @@ contract MixinTransactions is
...
+        if (tx.gasprice != requiredGasPrice) {
+            LibRichErrors.rrevert(LibExchangeRichErrors.TransactionGasPriceError(
+                transactionHash,
+                tx.gasprice,
+                requiredGasPrice
+            ));
+        }
+
         // Prevent reentrancy
         if (currentContextAddress != address(0)) {
```
Comment (jalextowle) [inline contracts/exchange/contracts/src/MixinTransactions.sol]: I think that there is a way for us to support nested `executeTransaction` calls. If we kept a stack variable called `startingContextAddress`, we could then replace our erasure of the `currentContextAddress` with the `startingContextAddress`. In this way, we'd get the best of both worlds by allowing users to submit metatransactions that submit other metatransactions (relayers could possibly make use of this), but they are always executed in the right context. Thoughts?

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+                actualGasPrice,
+                transaction.gasPrice,
+            );
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction, randomSignature(), {
+                gasPrice: actualGasPrice,
+            });
+            return expect(tx).to.revertWith(expectedError);
+        });
+        // FIXME - This should be unskipped when the contracts have been updated to fix this problem.
+        it.skip('should revert if reentrancy occurs in the middle of an executeTransaction call and msg.sender == signer for both calls', async () => {
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/exchange/test/transactions_unit_tests.ts]: This test will be removed.

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+                transactionHash1,
+            ]);
+            const expectedError = new ExchangeRevertErrors.TransactionExecutionError(transactionHash2, errorData);
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction2, validSignature, {
+                from: accounts[1], // Different then the signing addresses
+            });
+            return expect(tx).to.revertWith(expectedError);
+        });
+        // FIXME - This should be unskipped when the contracts have been updated to fix this problem.
+        it.skip('should revert if reentrancy occurs in the middle of an executeTransaction call and msg.sender == signer and then msg.sender != sender', async () => {
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/exchange/test/transactions_unit_tests.ts]: This test will also be removed (it should fail).

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+        function getExecuteTransactionCallData(transaction: ZeroExTransaction, signature: string): string {
+            return (transactionsContract as any).executeTransaction.getABIEncodedTransactionData(
+                transaction,
+                signature,
+            );
+        }
+        it('should revert if the current time is past the expiration time', async () => {
+            const currentTimestamp = await getLatestBlockTimestampAsync();
+            const transaction = await generateZeroExTransactionAsync({
+                expirationTimeSeconds: new BigNumber(currentTimestamp).minus(10), // Set the expiration time to before the current timestamp
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/exchange/test/transactions_unit_tests.ts]: I don't have a strong preference. We aren't using it a ton right now, so it might not be worth the refactor.

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 5]
```diff
@@ -67,23 +68,53 @@ blockchainTests.resets.only('Transaction Unit Tests', ({ provider, web3Wrapper,
...
+    ): Promise<ZeroExTransaction> {
+        const shouldSucceed = opts.shouldSucceed === undefined ? true : opts.shouldSucceed;
+        const callData = opts.callData === undefined ? constants.NULL_BYTES : opts.callData;
+        const returnData = opts.returnData === undefined ? constants.NULL_BYTES : opts.returnData;
+        const data = opts.data === undefined ? getExecutableCallData(shouldSucceed, callData, returnData) : opts.data;
+        const gasPrice = opts.gasPrice === undefined ? new BigNumber(constants.DEFAULT_GAS_PRICE) : opts.gasPrice;
+        const _domain = opts.domain === undefined ? domain : opts.domain;
+        const expirationTimeSeconds =
+            opts.expirationTimeSeconds === undefined
+                ? new BigNumber(await getLatestBlockTimestampAsync()).plus(10)
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/exchange/test/transactions_unit_tests.ts]: I guess we could just use `0` and `MAX_UINT256` for all of these values. I can make that change.

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+
+            // Create an abiEncoder for bytes. This will be used to decode the result and encode what
+            // is expected.
+            const abiEncoder = AbiEncoder.create('bytes');
+
+            // Ensure that the result contains the abi-encoded bytes "0xdeadbeef"
+            const encodedDeadbeef = abiEncoder.encode('0xdeadbeef');
+            expect(
+                result[0] ===
+                    '0x0000000000000000000000000000000000000000000000000000000000000020'.concat(
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/exchange/test/transactions_unit_tests.ts]: Found a super clean solution to this :)

File: packages/order-utils/src/exchange_revert_errors.ts [hunk 7]
```diff
@@ -207,6 +207,20 @@ export class TransactionExecutionError extends RevertError {
     }
 }
 
+export class TransactionGasPriceError extends RevertError {
+    constructor(transactionHash?: string, actualGasPrice?: BigNumber, requiredGasPrice?: BigNumber) {
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply packages/order-utils/src/exchange_revert_errors.ts]: We just use `BigNumber` in more of the constructors than not, so I will remove the `number | string` in any other constructors.

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+            const abiEncoder = AbiEncoder.createMethod('TransactionError', ['uint8', 'bytes32']);
+            const errorData = abiEncoder.encode([
+                ExchangeRevertErrors.TransactionErrorCode.NoReentrancy,
+                transactionHash1,
+            ]);
+            const expectedError = new ExchangeRevertErrors.TransactionExecutionError(transactionHash2, errorData);
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction2, validSignature, {
+                from: accounts[0],
+            });
+            return expect(tx).to.revertWith(expectedError);
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/exchange/test/transactions_unit_tests.ts]: Do we want to test for that or just remove the test?

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+            expect(logs[0].args.returnData).to.be.eq('0xdeadbeef');
+            expect(logs[1].event).to.be.eq('TransactionExecution');
+            expect(logs[1].args.transactionHash).to.eq(transactionHash1);
+            expect(logs[2].event).to.be.eq('ExecutableCalled');
+            expect(logs[2].args.data).to.be.eq(constants.NULL_BYTES);
+            expect(logs[2].args.returnData).to.be.eq('0xbeefdead');
+            expect(logs[3].event).to.be.eq('TransactionExecution');
+            expect(logs[3].args.transactionHash).to.eq(transactionHash2);
+        });
+    });
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/exchange/test/transactions_unit_tests.ts]: This should pass. Why is it an edge case?

[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/exchange/test/transactions_unit_tests.ts]: Because it's an ultra specific scenario that takes advantage of all the reentrancy and context management features in the system. Basically, if we changed any one of them this would probably fail so it might be a good way to check for regressions.

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 10]
```diff
@@ -357,99 +361,81 @@ blockchainTests.resets('Transaction Unit Tests', ({ provider, web3Wrapper, txDef
...
-            const expectedError = new ExchangeRevertErrors.TransactionExecutionError(transactionHash2, errorData);
-            const tx = transactionsContract.executeTransaction.sendTransactionAsync(transaction2, validSignature, {
-                from: accounts[0],
+            const outerTransactionHash = transactionHashUtils.getTransactionHashHex(outerTransaction);
+            const errorData = new ExchangeRevertErrors.TransactionInvalidContextError(
+                innerTransactionHash,
+                accounts[0],
+            ).encode();
+            const expectedError = new ExchangeRevertErrors.TransactionExecutionError(outerTransactionHash, errorData);
+            const tx = transactionsContract.executeTransaction.sendTransactionAsync(outerTransaction, validSignature, {
```
Comment (dorothy-zbornak) [inline contracts/exchange/test/transactions_unit_tests.ts]: These should all be `awaitTransactionSuccessAsync()` calls, even if they revert.

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 11]
```diff
@@ -0,0 +1,677 @@
...
+            const transactionHash = transactionHashUtils.getTransactionHashHex(transaction);
+
+            const [result, receipt] = await transactionHelper.getResultAndReceiptAsync(
+                transactionsContract.executeTransaction,
+                transaction,
+                validSignature,
+                { from: accounts[0] },
+            );
+
+            expect(transactionsContract.executeTransaction.getABIDecodedReturnData(result)).to.equal(
```
Comment (jalextowle) [inline contracts/exchange/test/transactions_unit_tests.ts]: 🎉

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 12]
```diff
@@ -0,0 +1,707 @@
...
+                accounts[0],
+                INVALID_SIGNATURE,
+            );
+            expect(
+                transactionsContract.assertExecutableTransaction.callAsync(transaction, INVALID_SIGNATURE, {
+                    from: accounts[1],
+                }),
+            ).to.revertWith(expectedError);
+        });
+        it('should be successful if signer == msg.sender and the signature is invalid', async () => {
```
Comment (jalextowle) [inline contracts/exchange/test/transactions_unit_tests.ts]: I think that we should also check that the `tx` should be successful if `signer == msg.sender and the signature is valid`. It's a bit pedantic, but it would make the suite more complete.

File: contracts/exchange/test/transactions_unit_tests.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (jalextowle) [reply contracts/exchange/test/transactions_unit_tests.ts]: @dorothy-zbornak Do you think it would make sense to add a simpler test for this? In this case, we could nest in an equally valid way by excluding the `batchExecuteTransaction` and just adding a single call to `executeTransaction`. Not to say that we shouldn't include the test case that you describe, but I think it would also be good to have a super simple case.

File: contracts/exchange/test/transactions_unit_tests.ts [hunk 13]
```diff
@@ -0,0 +1,707 @@
...
+                ExchangeRevertErrors.TransactionErrorCode.Expired,
+                transactionHash,
+            );
+            const tx = transactionsContract.executeTransaction.awaitTransactionSuccessAsync(
+                transaction,
+                randomSignature(),
+            );
+            return expect(tx).to.revertWith(expectedError);
+        });
+        it('should revert if the transaction is submitted with a gasPrice that does not equal the required gasPrice', async () => {
```
Comment (jalextowle) [inline contracts/exchange/test/transactions_unit_tests.ts]: Since you already test this behavior in `assertExecutableTransaction`'s test, I'd be fine if we want to remove this. Any reason why we shouldn't?

File: contracts/exchange/test/transactions_unit_tests.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/exchange/test/transactions_unit_tests.ts]: Yeah, we could probably afford to peel off one layer. I think all the other stuff kind of has to go together to hit all the paths at once, though. :thinking: I do think we're pretty well covered for now.

File: contracts/exchange/test/transactions_unit_tests.ts (hunk 13 above)
Comment (abandeali1) [reply contracts/exchange/test/transactions_unit_tests.ts]: I feel like most of the other tests are actually covered by the `assertExecutableTransaction` tests. We have extra tests for expired transactions, valid/invalid signatures, already executed transactions, etc. If we remove this, it probably makes sense to remove the other redundant tests as well.

Comment (jalextowle) [reply contracts/exchange/test/transactions_unit_tests.ts]: I guess out of these options, it might make sense to just keep them to prevent regressions. Longer term, I do think we should try to reduce test duplication, but it's easy to understand which are duplicates in this file.
